### PR TITLE
[GEOS-10222] WFS CSV OutputFormat delimiter

### PR DESCRIPTION
**Pull request by shadab-geo**
[GEOS-10222](https://osgeo-org.atlassian.net/browse/GEOS-10222)

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Review by mprins — CHANGES_REQUESTED** [review]
LGTM, but since this is a user visible change documentation must be updated in https://github.com/geoserver/geoserver/blob/main/doc/en/user/source/services/wfs/outputformats.rst

File: doc/en/user/source/services/wfs/outputformats.rst [hunk 1]
```diff
@@ -135,3 +135,15 @@ JSON output ``format_options``:
...
+
+
+CSV output
+----------------
+
+A Default CSV file uses a comma to separate values. Each line of the file is a data record. Each record consists of one or more fields, separated by commas. The separator can be changed using format_options as specified below.
+
+csv file output ``format_options``:
+
+* ``format_option=filename:<file>``: if a file name is provided, the name is used as the output file name. For example, ``format_options=filename:roads.zip``.
```
Comment (mprins) [inline doc/en/user/source/services/wfs/outputformats.rst:147]: even though correct, I find it a bit counter-intuitive to have a CSV file with `.zip` extension
```suggestion
* ``format_option=filename:<file>``: if a file name is provided, the name is used as the output file name. For example, ``format_options=filename:roads.csv``.
```

**Review by mprins — APPROVED** [review]
thanks for the docs, one small nit-pick

File: doc/en/user/source/services/wfs/outputformats.rst (hunk 1 above)
Comment (shadab-geo) [reply doc/en/user/source/services/wfs/outputformats.rst]: Changes applied, Thank you :)

**Comment by jratike80** [comment]
If I read the code right, the separator character is somehow escaped if it appears in a text string. How does it alter for example string `osm:highway` when `:` is used as a separator?

**Comment by shadab-geo** [comment]
> If I read the code right, the separator character is somehow escaped if it appears in a text string. How does it alter for example string `osm:highway` when `:` is used as a separator?

It will enclose the string with double quotes and return it as `"osm:highway"`

**Comment by jratike80** [comment]
All right. What if source data is like

```
osm:highway
osm_highway
```

will the output CSV with : as separator be like

```
attr:"osm:highway":next_attr
atrr:osm_highway:next_attr
```
I can see from topp:states
http://localhost:8080/geoserver/topp/ows?service=WFS&version=1.0.0&request=GetFeature&typeName=topp%3Astates3&maxFeatures=50&outputFormat=text%2Fcsv
that by default the geometry WKT is enclosed between "" but the string attributes are not:
`Missouri,29,W N Cen,MO,`

I wonder if it would be better to enclose always all the strings between double quotes
http://localhost:8080/geoserver/ows?service=wfs&version=2.0.0&request=DescribeFeatureType&typenames=topp:states
->
`"Missouri","29","W N Cen","MO",`

Pay attention to state_fips that is string in the schema but that has value 29. CSV readers may well interpret the attribute to be numeric.

Additional thinking is needed if user provides `"` as a separator (action: deny?), or strings contain " characters (action: escape \\" ?).

**Comment by shadab-geo** [comment]
> All right. What if source data is like
> 
> ```
> osm:highway
> osm_highway
> ```
> 
> will the output CSV with : as separator be like
> 
> ```
> attr:"osm:highway":next_attr
> atrr:osm_highway:next_attr
> ```
That's right!

> I can see from topp:states
> http://localhost:8080/geoserver/topp/ows?service=WFS&version=1.0.0&request=GetFeature&typeName=topp%3Astates3&maxFeatures=50&outputFormat=text%2Fcsv
> that by default the geometry WKT is enclosed between "" but the string attributes are not:
> `Missouri,29,W N Cen,MO,`

This is expected behavior. The CSV separator, in this case, is a comma. The values of column 'the_geom' contain a comma in its value which is why it encloses it within the double-quotes.

> I wonder if it would be better to enclose always all the strings between double quotes
> http://localhost:8080/geoserver/ows?service=wfs&version=2.0.0&request=DescribeFeatureType&typenames=topp:states
> ->
> `"Missouri","29","W N Cen","MO",`

I am not sure about it. We need to check with other devs because this logic of enclosing string within double quote was already existing. (I have just added dynamic CSV separator in this PR)

> Pay attention to state_fips that is string in the schema but that has value 29. CSV readers may well interpret the attribute to be numeric.

consider a case with a negative number and dash as a separator, I guess it will crash the CSV format in that case.

> Additional thinking is needed if user provides `"` as a separator (action: deny?), or strings contain " characters (action: escape \" ?).

I agree we might want to deny adding `"` as a separator. 
If string contains `"`  for example `Mi"ssouri,29`
it will write as
` "Mi"ssouri",29`

**Comment by dromagnoli** [comment]
@shadab-geo any other action needed on this ticket?

**Comment by shadab-geo** [comment]
Adding fix to prevent double-quotes @dromagnoli

**Pull request merged**